File: Java/Language/Generics/TypeParamTest2/SimpleStack.java
//a generic class declaration with type-parameter V
class SimpleStack<V> {

	private Node top;

	//inner (non-static) member class can refer to any member 
	//of outer class
	class Node {

		V value;
		Node below;

		Node(V item) {
			value = item;
			below = top;
		}
	}

	public void push(V item) {
		top = new Node(item);
	}

	public V pop() {
		V item = top.value;
		top = top.below;
		return item;
	}

	public boolean empty() {
		return top == null;
	}
}

File: dotnet/Language/Generics/TypeParamTest2/DemoApp/SimpleStack.cs
//covariant interface over V
interface IStackReader<out V>
{
    V Pop();
    bool Empty();
}

//contra-variant interface over V
interface IStackWriter<in V>
{
    void Push(V item);
}

class SimpleStack<V> : IStackReader<V>, IStackWriter<V>
{
    private Node top;

    class Node
    {
        internal V value;
        internal Node below;

        internal Node(V value, Node below)
        {
            this.value = value;
            this.below = below;
        }
    }

    public void Push(V item)
    {
        top = new Node(item, top);
    }

    public V Pop()
    {
        V item = top.value;
        top = top.below;
        return item;
    }

    public bool Empty()
    {
        return top is null; //top == null
    }
}